### PR TITLE
qemu_monitor.py: Add bitmap and live backup command.

### PR DESCRIPTION
**Pull request by vivianQizhu**
Add block backup support.

Signed-off-by: Qianqian Zhu <qizhu@redhat.com>

### CODE REVIEW

File: virttest/qemu_monitor.py [hunk 1]
```diff
@@ -2373,3 +2373,78 @@ def transaction(self, job_list):
         """
         transaction_args = {"actions": job_list}
         return self.cmd("transaction", transaction_args)
+
+    def add_dirty_bitmap(self, node, name, granularity=65536):
```
Comment (balamuruhans) [inline virttest/qemu_monitor.py]: methods add_dirty_bitmap(), remove_dirty_bitmap(), clear_dirty_bitmap seems to be similar, just curious to know do we really need to have separate methods.
Instead,

```
def dirty_bitmap(self, operation, node, name, granularity=65536):
    cmd = "block-dirty-bitmap-%s" % (operation)
    self.verify_supported_cmd(cmd)
    args = {"node": node, "name": name}
    if granularity:
        args["granularity"] = granularity
    return self.cmd(cmd, args)
```

Comment (vivianQizhu) [inline virttest/qemu_monitor.py]: @balamuruhans thanks, yours looks better, will update.

**Comment by vivianQizhu** [comment]
22:44:22 DEBUG| (monitor qmp1) Sending command 'block-dirty-bitmap-add'
22:44:22 DEBUG| Send command: {'execute': 'block-dirty-bitmap-add', 'arguments': {'node': u'drive_image1', 'name': 'bitmap0', 'granularity': 32768}, 'id': 'KExwSKM6'}
22:44:22 INFO | Create full backup for drive_image1
22:44:22 DEBUG| (monitor qmp1) Sending command 'drive-backup'
22:44:22 DEBUG| Send command: {'execute': 'drive-backup', 'arguments': {'device': u'drive_image1', 'mode': 'absolute-paths', 'sync': 'full', 'target': '/usr/share/avocado/data/avocado-vt/images/fullbackup.qcow2', 'format': 'qcow2'}, 'id': 'J2vFiPcY'}

22:47:55 INFO | Create incremental backup for drive_image1
22:47:55 DEBUG| (monitor qmp1) Sending command 'drive-backup'
22:47:55 DEBUG| Send command: {'execute': 'drive-backup', 'arguments': {'target': '/usr/share/avocado/data/avocado-vt/images/incremental0.qcow2', 'format': 'qcow2', 'sync': 'incremental', 'bitmap': 'bitmap0', 'mode': 'existing', 'device': u'drive_image1'}, 'id': 'edkn5Mog'}

22:49:09 INFO | PASS 1-smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.myprovider.live_backup.granularity.32768

**Comment by vivianQizhu** [comment]
@balamuruhans Updated, would you help review it again? Thanks.

File: virttest/qemu_monitor.py [hunk 2]
```diff
@@ -2373,3 +2373,50 @@ def transaction(self, job_list):
...
+
+        :param operation: operations to bitmap, can be 'add', 'remove', and 'clear'
+        :param node: device/node on which to operate dirty bitmap
+        :param name: name of the dirty bitmap to operate
+        :param granularity: granularity to track writes with
+        """
+        cmd = "block-dirty-bitmap-%s" % operation
+        self.verify_supported_cmd(cmd)
+        args = {"node": node, "name": name}
+        if operation == "add":
```
Comment (balamuruhans) [inline virttest/qemu_monitor.py]: its a good practise to compare strings with lower() or upper(), if operation.lower() == "add":

File: virttest/qemu_monitor.py [hunk 3]
```diff
@@ -2373,3 +2373,50 @@ def transaction(self, job_list):
...
+            (NewImageMode, optional, default 'absolute-paths')
+        """
+        cmd = "drive-backup"
+        self.verify_supported_cmd(cmd)
+        args = {"device": device,
+                "target": target,
+                "format": format,
+                "sync": sync,
+                "mode": mode}
+        if sync == "incremental":
```
Comment (balamuruhans) [inline virttest/qemu_monitor.py]: same here.

File: virttest/qemu_monitor.py [hunk 4]
```diff
@@ -2373,3 +2373,50 @@ def transaction(self, job_list):
...
+        :param granularity: granularity to track writes with
+        """
+        cmd = "block-dirty-bitmap-%s" % operation
+        self.verify_supported_cmd(cmd)
+        args = {"node": node, "name": name}
+        if operation == "add":
+            args["granularity"] = granularity
+        return self.cmd(cmd, args)
+
+    def drive_backup(self, device, target, format, sync, speed=0,
```
Comment (balamuruhans) [inline virttest/qemu_monitor.py]: doc string for speed can also be added.

File: virttest/qemu_monitor.py (hunk 3 above)
Comment (vivianQizhu) [inline virttest/qemu_monitor.py]: @balamuruhans Thanks, updated.

**Comment by balamuruhans** [comment]
LGTM

File: virttest/qemu_monitor.py [hunk 5]
```diff
@@ -2373,3 +2373,51 @@ def transaction(self, job_list):
...
+
+        :param operation: operations to bitmap, can be 'add', 'remove', and 'clear'
+        :param node: device/node on which to operate dirty bitmap
+        :param name: name of the dirty bitmap to operate
+        :param granularity: granularity to track writes with
+        """
+        cmd = "block-dirty-bitmap-%s" % operation
+        self.verify_supported_cmd(cmd)
+        args = {"node": node, "name": name}
+        if operation.lower() == "add":
```
Comment (xutian) [inline virttest/qemu_monitor.py]: if  operation not in [ add, remove, create] exception will raised in line verify_support_cmd, so covert case here is useless.

File: virttest/qemu_monitor.py [hunk 6]
```diff
@@ -2373,3 +2373,51 @@ def transaction(self, job_list):
         """
         transaction_args = {"actions": job_list}
         return self.cmd("transaction", transaction_args)
+
+    def dirty_bitmap(self, operation, node, name, granularity=65536):
```
Comment (xutian) [inline virttest/qemu_monitor.py]: how about operate_dirty_bit_map()?

**Comment by vivianQizhu** [comment]
@xutian Thanks, updated. Would you help review it again?

**Comment by vivianQizhu** [comment]
@xiywang , Would you please help review this? Thanks.

**Comment by xiywang** [comment]
LGTM

**Comment by xutian** [comment]
Update your commit summary like:
qemu_monitor: add block back support
execute below command in your branch
git commit --amend 
then push your commit  with '--force' option

**Comment by vivianQizhu** [comment]
@xutian , thanks, updated

File: virttest/qemu_monitor.py (hunk 2 above)
Comment (xutian) [inline virttest/qemu_monitor.py]: @balamuruhans  if operation is not  low case, next line verify_supported_cmd(cmd) will raise exception so I think above check is unnecessary.

Thanks,
Xu

**Pull request merged**